File: class32_dp/CoinGame.java
import java.util.*;

public class CoinGame {

    static int coinGame(int[] coins) {
        int[][] strg = new int[coins.length][coins.length];

        for (int g = 0; g < strg.length; g++) {
            for (int i = 0; i < strg.length - g; i++) {
                int j = i + g;

                if (g == 0) {
                    strg[i][j] = coins[j];
                } else if (g == 1) {
                    int left = strg[i][j - 1];
                    int right = strg[i + 1][j];
                    strg[i][j] = Math.max(left, right);
                } else {
                    // The opponent will try to maximize his value
                    // so we'll try to account for that
                    // murphy's law
                    int min1 = Math.min(strg[i][j - 2], strg[i + 1][j - 1]);
                    int min2 = Math.min(strg[i + 2][j], strg[i + 1][j - 1]);

                    if (min1 + coins[j] < min2 + coins[i]) {
                        strg[i][j] = min2 + coins[i];
                    } else  {
                        strg[i][j] = min1 + coins[j];
                    }
                }
            }

            if (g != 0) {
                g++;
            }
        }

        for (int i = 0; i < strg.length; i++) {
            for (int j = 0; j < strg.length; j++) {
                System.out.printf("%3d", strg[i][j]);
            }
            System.out.println();
        }

        return strg[0][strg[0].length - 1];
    }

    public static void main(String[] args) {
        int[] coins = { 20, 30, 2, 2 };
        System.out.println(coinGame(coins));
    }
}

File: class32_dp/coin_game.cpp
#include <iostream>
#include <vector>
#include <chrono>
using namespace std;

int CoinGame(vector<int> &coins, vector<vector<int>> &strg, int i, int j)
{
    if (i == j)
    {
        return coins[i];
    }
    if (i > j)
    {
        return 0;
    }
    if (strg[i][j] != 0)
    {
        return strg[i][j];
    }

    // I choose first coin
    // -> then Player chooses second coin
    int val1 = CoinGame(coins, strg, i + 2, j);
    // -> or player chooses last coin
    int val2 = CoinGame(coins, strg, i + 1, j - 1);

    // Player tries to maximize his score by taking max of val1 and val2
    // I am left with the minimum of val1 and val2
    int ret1 = min(val1, val2);

    // I choose second coin
    // -> then Player chooses first coin
    int val3 = CoinGame(coins, strg, i + 1, j - 1);
    // -> or Player chooses second last coin
    int val4 = CoinGame(coins, strg, i, j - 2);

    // Player tries to maximize his score by taking max of val3 and val4
    // I am left with the minimum of val3 and val4
    int ret2 = min(val3, val4);

    // I try to maximize my score by choosing that coin which gives me best solution
    // I take maximum of coin 1 + ret1 and coin2 + ret2
    int bestTotal = max(coins[i] + ret1, coins[j] + ret2);
    strg[i][j] = bestTotal;
    return bestTotal;
}

int main(int argc, char **argv)
{
    vector<int> coins({20, 30, 2, 2, 2, 10});
    vector<vector<int>> strg(coins.size(), vector<int>(coins.size(), 0));
    auto start = chrono::high_resolution_clock::now();
    cout << CoinGame(coins, strg, 0, coins.size() - 1) << endl;
    auto end = chrono::high_resolution_clock::now();
    cout << "Time taken " << chrono::duration_cast<chrono::milliseconds>(end - start).count() << " ms" << endl;
}
